--- PKGBUILD	2022-01-07 00:16:16.761634963 +0300
+++ PKGBUILD	2022-01-07 00:16:20.298307826 +0300
@@ -6,7 +6,7 @@
 pkgdesc="Test building a Rust package with LTO enabled"
 arch=('x86_64')
 depends=('gcc-libs')
-makedepends=('cargo' 'git')
+makedepends=('cargo' 'git' 'clang' 'lld')
 source=("$pkgname::git+file:///$PWD/../app/")
 sha256sums=('SKIP')
 options=('lto')
@@ -18,6 +18,26 @@
 
 build() {
   cd "$pkgname"
+
+  if [[ $CFLAGS = *-flto* ]]; then
+    local rustflags=(
+      -C linker=clang
+      -C linker-plugin-lto
+      -C link-args="$CFLAGS -fuse-ld=lld"
+    )
+
+    IFS=$'\x1f'
+    export CARGO_ENCODED_RUSTFLAGS="${rustflags[*]}"
+    unset IFS
+
+    export CARGO_PROFILE_RELEASE_CODEGEN_UNITS=1
+    export CARGO_PROFILE_RELEASE_LTO=fat
+    export CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=clang
+
+    export CC=clang
+    export CXX=clang++
+  fi
+
   cargo build -vv --release --frozen
 }
 
